controller/src/main/java/io/pravega/controller/store/kvtable/AbstractTableMetadataStore.java

```diff
@@ -0,0 +1,207 @@
+/**
+ * Copyright (c) Dell Inc., or its subsidiaries. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ */
+package io.pravega.controller.store.kvtable;
+
+import com.google.common.cache.CacheBuilder;
+import com.google.common.cache.CacheLoader;
+import com.google.common.cache.LoadingCache;
+import io.pravega.client.tables.KeyValueTableConfiguration;
+import io.pravega.common.concurrent.Futures;
+import io.pravega.controller.store.Scope;
+import io.pravega.controller.store.OperationContext;
+import io.pravega.controller.store.VersionedMetadata;
+import io.pravega.controller.store.Artifact;
+import io.pravega.controller.store.index.HostIndex;
+import io.pravega.controller.store.stream.StoreException;
+import io.pravega.shared.controller.event.ControllerEvent;
+import io.pravega.shared.controller.event.ControllerEventSerializer;
+import org.apache.commons.lang3.tuple.ImmutablePair;
+import org.apache.commons.lang3.tuple.Pair;
+
+import javax.annotation.ParametersAreNonnullByDefault;
+import java.util.UUID;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.Executor;
+import java.util.concurrent.TimeUnit;
+
+public abstract class AbstractTableMetadataStore implements TableMetadataStore {
+    private final LoadingCache<String, Scope> scopeCache;
+    private final LoadingCache<Pair<String, String>, KeyValueTable> cache;
+    private final HostIndex hostTaskIndex;
+    private final ControllerEventSerializer controllerEventSerializer;
+
+    protected AbstractTableMetadataStore(HostIndex hostTaskIndex) {
+        cache = CacheBuilder.newBuilder()
+                .maximumSize(10000)
+                .refreshAfterWrite(10, TimeUnit.MINUTES)
+                .expireAfterWrite(10, TimeUnit.MINUTES)
+                .build(
+                        new CacheLoader<Pair<String, String>, KeyValueTable>() {
+                            @Override
+                            @ParametersAreNonnullByDefault
+                            public KeyValueTable load(Pair<String, String> input) {
+                                try {
+                                    return newKeyValueTable(input.getKey(), input.getValue());
+                                } catch (Exception e) {
+                                    throw new RuntimeException(e);
+                                }
+                            }
+                        });
+
+        scopeCache = CacheBuilder.newBuilder()
```

Comment:
we can take out the scope's object cache into its own class.. and use that same cache here and in stream metadata store.

Reply:
Will implement optimizations after the basic flow is all done and working.